File: neetcode/110. Balanced Binary Tree/110.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    boolean ans;
    public boolean isBalanced(TreeNode root) {
        if (root == null) return true;

        int leftHeight = findHeight(root.left);
        int rightHeight = findHeight(root.right);

        if (Math.abs(leftHeight - rightHeight) > 1) return false;

        return isBalanced(root.left) && isBalanced(root.right);

    }

    private int findHeight(TreeNode node) {
        if (node == null) return 0;

        int left = findHeight(node.left);
        int right = findHeight(node.right);

        return 1 + Math.max(left, right);
    }
}

File: neetcode/110. Balanced Binary Tree/110.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun isBalanced(root: TreeNode?): Boolean {
        if (root == null) return true

        val leftHeight = findHeight(root.left)
        val rightHeight = findHeight(root.right)

        if (abs(leftHeight - rightHeight) > 1) return false

        return isBalanced(root.left) && isBalanced(root.right)
    }

    fun findHeight(node: TreeNode?): Int {
        if (node == null) return 0

        val leftHeight = findHeight(node.left)
        val rightHeight = findHeight(node.right)

        return 1 + max(leftHeight, rightHeight)
    }
}
